### ISSUE TITLE
Iterating over a map value without NodeRef's .children() (indexing)

### ISSUE THREAD

**Issue by KadeDev** [biojppm/rapidyaml#351]
I've seen that indexing is the fastest way to get data from a tree, but I am trying to iterate over a map inside of a tree (using indexing, instead of NodeRef .children() etc as it seems to take forever on super long maps)

My main issue is, I can't seem to find a way to actually do this, and I was wondering if anyone could help me with that?

The project I am trying to use ryml on has user created content, so speed is very important to me.

**Comment by biojppm** [comment]
Have a look in the quickstart sample in the main README file. It has examples of iteration both with nodes and indices.

But you should be aware the speed difference is only a factor. Their algorithmic complexity is the same, so if you say one of them is slow the other one will also be slow. I suspect you're chaining [] operators in the body of the loop? Every [] call is a linear search, so if you're doing it in the body of the loop, the overall complexity is quadratic and then it can get slow.

Otherwise, I will need an example of your code and data to figure why it may be slow.

**Comment by KadeDev** [comment]
Yeah I dont know if this would be even a speed improvement over simply using noderef.

My main idea looking is using the first_child(id) with the id being the map's id, and then using the tree.find_child(id, "property").

But it seems to be the same speed as noderef (possibly slower).

My fastest code was
```cpp
ryml::ConstNodeRef root = tree.rootref();

for(ryml::ConstNodeRef const& child : root["TimingPoints"].children()) // roughly 21,000 maps
{
    TimingPoint p;
    // ... code to create a timing point etc, referencing properties like child["StartTime"], and child["Bpm"] (converting them to floats)
}
```

Taking around 890ms.

I do believe what you are saying about chaining [] could be an issue, but I dont think I am doing that recklessly.

**Comment by KadeDev** [comment]
The data set I am using can be found here (be in mind, yes it is extreme. Redundant, but due to my user generated content based project. It is an inevitability that this would exist somewhere.) https://pastebin.com/qp9JhQq3

**Comment by biojppm** [comment]
I need to see the code to create your `TimingPoint`. It's hard to know what's going on otherwise.

**Comment by KadeDev** [comment]
> I need to see the code to create your `TimingPoint`. It's hard to know what's going on otherwise.

I dont have the code on hand but I did describe it in the comment.

I'll try to write it again, I am pretty sure it went something like this:

```cpp
TimingPoint p;

if (child.has_child("StartTime"))
    p.StartTime = std::stof(child["StartTime"].val().str);
else
    p.StartTime = -1;
p.Bpm = std::stof(child["Bpm"].val().str);
```

Is this maybe because of the "has_child" or the ".str"?

**Comment by NaN-git** [comment]
> ```c++
> TimingPoint p;
> 
> if (child.has_child("StartTime"))
>     p.StartTime = std::stof(child["StartTime"].val().str);
> else
>     p.StartTime = -1;
> p.Bpm = std::stof(child["Bpm"].val().str);
> 
> points.push_back(p);
> ```
Please, don't access member variables if possible. Your main problem is that `val().str` returns `char *`, which is implicitly converted to `std::string` before `std::stof` is called. But ryml does not use C-strings (zero terminated) and your whole file will be copied over and over again by the constructor of `std::string`.
Another point for optimization is the use of `has_child` and later accessing the child with `operator[]` because both functions involve a linear search.
There is a builtin function to get `StartTime` as float:
```
float time;
child.get_if("StartTime", &time, -1.f);
```

**Comment by biojppm** [comment]
Thanks for pointing it out, @NaN-git . @KadeDev, the use of stof is indeed the problem with your code. It will read to the end of the file on every call, because the strings are not zero terminated. So you have a quadratic loop with a very large N. And if the full string contents are not zero terminated, this may very well overflow the parsed buffer.

To clarify further, accessing the member variable of the substr is fine IF you are aware that it is not zero terminated.

Also, after fixing the issues above, you may gain even more by iterating the child's children in order and switching on their key. So even the call to `.get_if()` is not needed. This will save you doing any linear search. I'm guessing `TimingPoint` does not have many members, so this will also mean a considerable further gain.

**Comment by biojppm** [comment]
To elaborate further:

* `csubstr` is NOT zero terminated, unlike C-strings or `std::string`. It's just a pointer and a length.
* `std::stof()` receives a `std::string`. Calling `std::stof(child["StartTime"].val().str)` will implicitly call the string constructor `std::string(const char*)`, causing the given pointer to be scanned to the end of the parse buffer. It will overflow the input string, and if there's no terminating `\0` at the end of the buffer it will also overflow the buffer. 
*  you can still use `std::stof()` if you call the proper constructor: `std::string(const char* s, size_t len)`. So you would do something like:
   ```c++
   TimingPoint p;

   if (child.has_child("StartTime")) // 1 linear search
   {
       csubstr val = child["StartTime"].val(); // 1 linear search
       p.StartTime = std::stof(val.str, val.len); // length explicitly given, no scan occurs
   }
   else
       p.StartTime = -1;
   p.Bpm = std::stof(child["Bpm"].val().str);  // 1 linear search
   // more members..., each one a linear search
   ```
* While the above code will no longer suffer from the quadratic behavior, it is still bad because it searches twice for `StartTime` (which is a linear search). Like @NaN-git points out, this will give you what you want, with a single search (notice how much more concise it is, too):
   ```c++
   TimingPoint p;

   child.get_if("StartTime", &p.StartTime, -1.f); // 1 linear search
   child["Bpm"] >> p.Bpm; // 1 linear search
   // more members..., each one a linear search
   ```
  so this only performs one search per member.
* And the final suggestion is the following. Instead of searching at all, why not just iterate?
   ```c++
   TimingPoint p;
   for(ConstNodeRef member : child.children())
   {
        csubstr key = member.key();
        if(key == "StartTime")
             member >> p.StartTime;
        else if(key == "Bpm")
             member >> p.Bpm;
        // etc
   }
   ```
   So this will perform only a single search. If the number of members to read for `TimingPoint` is more than a handful, this will probably be a net performance gain.

**Comment by KadeDev** [comment]
Ah I see, I was using this library wrong.

Thank you for pointing this out!

After taking this into account, it is much faster and a great learning experience.
